File: c/ch5/poisson.c

static char help[] = "Solves a structured-grid Poisson problem with DMDA and KSP,\n"
"but (unlike c2poisson) also using KSPSetComputeOperators() so we can use\n"
"multigrid preconditioning at the command line.\n\n";

// based on src/ksp/ksp/examples/tutorials/ex50.c, but in Dirichlet-only case,
//   and following c2poisson.c closely

// SHOWS c2 AND c4 CODES ARE DOING SAME THING
//   $ ./c2poisson
//   on 10 x 10 grid:  error |u-uexact|_inf = 0.000621778
//   $ ./c4poisson 
//   on 10 x 10 grid:  error |u-uexact|_inf = 0.000621778

// REGARDLESS OF INTEGER DIMS OF INITIAL GRID, THIS ALWAYS WORKS:
//   ./c4poisson -pc_type mg -pc_mg_levels N -da_refine N
// SO CONVERGENCE IS:
//   for NN in 1 2 3 4 5 6 7 8; do ./c4poisson -da_grid_x 5 -da_grid_y 5 -da_refine $NN -pc_type mg -pc_mg_levels $NN -ksp_rtol 1.0e-8; done

// USE MULTIGRID AND SHOW IT GRAPHICALLY:
//   ./c4poisson -da_grid_x 3 -da_grid_y 3 -pc_type mg -da_refine 3 -ksp_monitor -ksp_view -dm_view draw -draw_pause 1

// PERFORMANCE ANALYSIS; COMPARE c2poisson VERSION:
//   export PETSC_ARCH=linux-gnu-opt
//   make c4poisson
//   ./c4poisson -da_grid_x 3 -da_grid_y 3 -pc_type mg -da_refine 9 -log_summary|grep "Solve: "
//   mpiexec -n 6 FIXME

// NOT SURE WHAT IS BEING ACCOMPLISHED HERE:
//   ./c4poisson -da_grid_x 100 -da_grid_y 100 -pc_type mg  -pc_mg_levels 1 -mg_levels_0_pc_type ilu -mg_levels_0_pc_factor_levels 1 -ksp_monitor -ksp_view
//   ./c4poisson -da_grid_x 100 -da_grid_y 100 -pc_type mg -pc_mg_levels 1 -mg_levels_0_pc_type lu -mg_levels_0_pc_factor_shift_type NONZERO -ksp_monitor

// I NEED TO KNOW/UNDERSTAND EVERYTHING IN foo.txt:
//   mpiexec -n 4 ./c4poisson -da_grid_x 3 -da_grid_y 3 -pc_type mg -da_refine 10 -ksp_monitor -dm_view -ksp_view -log_summary &> foo.txt

// ONLY 17 SECONDS BUT USES 9 GB MEMORY:
//   mpiexec -n 4 ./c4poisson -da_grid_x 3 -da_grid_y 3 -pc_type mg -da_refine 11 -ksp_monitor
// (COMPARE: mpiexec -n 4 ./c2poisson -da_grid_x 4097 -da_grid_y 4097 -ksp_type cg
// WHICH DOES 3329 ITERATIONS)

// COMPARABLE?:
//mpiexec -n 4 ./c4poisson -da_grid_x 4097 -da_grid_y 4097 -pc_type mg -pc_mg_levels 11 -ksp_monitor -pc_mg_type full -ksp_rtol 1.0e-12 -mg_levels_ksp_type cg

#include <petsc.h>
#include "../ch3/structuredpoisson.h"

//CONTEXT
typedef struct {
  DMDALocalInfo info;
  PetscReal     hx, hy;
} PoissonCtx;
//ENDCONTEXT

//COMPUTES
PetscErrorCode ComputeRHS(KSP ksp, Vec b, void *ctx) {
  PetscErrorCode ierr;
  DM             da;
  PoissonCtx     *pctx = (PoissonCtx*)ctx;
  PetscFunctionBeginUser;
  ierr = KSPGetDM(ksp,&da);CHKERRQ(ierr);
  ierr = formRHS(da,pctx->info,pctx->hx,pctx->hy,b); CHKERRQ(ierr);
  return 0;
}

PetscErrorCode ComputeA(KSP ksp, Mat J, Mat A, void *ctx) {
  PetscErrorCode ierr;
  DM             da;
  PoissonCtx     *pctx = (PoissonCtx*)ctx;
  PetscFunctionBeginUser;
  ierr = KSPGetDM(ksp,&da); CHKERRQ(ierr);
  ierr = formdirichletlaplacian(da,pctx->info,pctx->hx,pctx->hy,1.0,A); CHKERRQ(ierr);
  PetscFunctionReturn(0);
}
//ENDCOMPUTES


//MAIN
int main(int argc,char **argv)
{
  PetscErrorCode ierr;
  PoissonCtx     ctx;
  PetscInitialize(&argc,&argv,(char*)0,help);

  DM da;
  ierr = DMDACreate2d(PETSC_COMM_WORLD,
                DM_BOUNDARY_NONE, DM_BOUNDARY_NONE, DMDA_STENCIL_STAR,
                -10,-10,PETSC_DECIDE,PETSC_DECIDE,1,1,NULL,NULL,
                &da); CHKERRQ(ierr);
  ierr = DMDASetUniformCoordinates(da,0.0,1.0,0.0,1.0,-1.0,-1.0); CHKERRQ(ierr);
  ierr = DMDAGetLocalInfo(da,&ctx.info); CHKERRQ(ierr);
  ctx.hx = 1.0/(ctx.info.mx-1);
  ctx.hy = 1.0/(ctx.info.my-1);

  // create linear solver context; compare to c2poisson.c version; note there
  // is no "Assemble" stage!; that happens inside KSPSolve()
  KSP ksp;
  ierr = KSPCreate(PETSC_COMM_WORLD,&ksp); CHKERRQ(ierr);
  ierr = KSPSetDM(ksp,da); CHKERRQ(ierr);
  ierr = KSPSetComputeRHS(ksp,ComputeRHS,&ctx); CHKERRQ(ierr);
  ierr = KSPSetComputeOperators(ksp,ComputeA,&ctx); CHKERRQ(ierr);
  ierr = KSPSetFromOptions(ksp); CHKERRQ(ierr);

  Vec u;
  ierr = DMCreateGlobalVector(da,&u); CHKERRQ(ierr);
  PetscLogStage  stage; //STRIP
  ierr = PetscLogStageRegister("Solve", &stage); CHKERRQ(ierr); //STRIP
  ierr = PetscLogStagePush(stage); CHKERRQ(ierr); //STRIP
  ierr = KSPSolve(ksp,NULL,u); CHKERRQ(ierr);
  ierr = PetscLogStagePop(); CHKERRQ(ierr); //STRIP

  PetscScalar    errnorm;
  Vec            uexact;
  ierr = DMCreateGlobalVector(da,&uexact); CHKERRQ(ierr);
  ierr = formExact(da,ctx.info,ctx.hx,ctx.hy,uexact); CHKERRQ(ierr);
  ierr = VecAXPY(u,-1.0,uexact); CHKERRQ(ierr);    // u <- u + (-1.0) uxact
  ierr = VecNorm(u,NORM_INFINITY,&errnorm); CHKERRQ(ierr);
  ierr = PetscPrintf(PETSC_COMM_WORLD,
             "on %d x %d grid:  error |u-uexact|_inf = %g\n",
             ctx.info.mx,ctx.info.my,errnorm); CHKERRQ(ierr);

  VecDestroy(&u);  VecDestroy(&uexact);
  KSPDestroy(&ksp);
  DMDestroy(&da);
  ierr = PetscFinalize(); CHKERRQ(ierr);
  return 0;
}
//ENDMAIN

File: python/ch3/poisson.py
#!/usr/bin/env python3
'''A structured-grid Poisson solver using DMDA+KSP.'''

import sys
import numpy as np
import petsc4py
petsc4py.init(sys.argv)  # must come before "import PETSc"
from petsc4py import PETSc

FIXME

#extern PetscErrorCode formMatrix(DM, Mat);
#extern PetscErrorCode formExact(DM, Vec);
#extern PetscErrorCode formRHS(DM, Vec);

#    DM            da;
#    Mat           A;
#    Vec           b,u,uexact;
#    KSP           ksp;
#    PetscReal     errnorm;
#    DMDALocalInfo info;


    // change default 9x9 size using -da_grid_x M -da_grid_y N
    ierr = DMDACreate2d(PETSC_COMM_WORLD,
                 DM_BOUNDARY_NONE, DM_BOUNDARY_NONE, DMDA_STENCIL_STAR,
                 9,9,PETSC_DECIDE,PETSC_DECIDE,1,1,NULL,NULL,&da); CHKERRQ(ierr);

    // create linear system matrix A
    ierr = DMSetFromOptions(da); CHKERRQ(ierr);
    ierr = DMSetUp(da); CHKERRQ(ierr);
    ierr = DMCreateMatrix(da,&A); CHKERRQ(ierr);
    ierr = MatSetFromOptions(A); CHKERRQ(ierr);

    // create RHS b, approx solution u, exact solution uexact
    ierr = DMCreateGlobalVector(da,&b); CHKERRQ(ierr);
    ierr = VecDuplicate(b,&u); CHKERRQ(ierr);
    ierr = VecDuplicate(b,&uexact); CHKERRQ(ierr);

    // fill vectors and assemble linear system
    ierr = formExact(da,uexact); CHKERRQ(ierr);
    ierr = formRHS(da,b); CHKERRQ(ierr);
    ierr = formMatrix(da,A); CHKERRQ(ierr);

    // create and solve the linear system
    ierr = KSPCreate(PETSC_COMM_WORLD,&ksp); CHKERRQ(ierr);
    ierr = KSPSetOperators(ksp,A,A); CHKERRQ(ierr);
    ierr = KSPSetFromOptions(ksp); CHKERRQ(ierr);
    ierr = KSPSolve(ksp,b,u); CHKERRQ(ierr);

    // report on grid and numerical error
    ierr = VecAXPY(u,-1.0,uexact); CHKERRQ(ierr);    // u <- u + (-1.0) uxact
    ierr = VecNorm(u,NORM_INFINITY,&errnorm); CHKERRQ(ierr);
    ierr = DMDAGetLocalInfo(da,&info);CHKERRQ(ierr);
    ierr = PetscPrintf(PETSC_COMM_WORLD,
                "on %d x %d grid:  error |u-uexact|_inf = %g\n",
                info.mx,info.my,errnorm); CHKERRQ(ierr);

    VecDestroy(&u);  VecDestroy(&uexact);  VecDestroy(&b);
    MatDestroy(&A);  KSPDestroy(&ksp);  DMDestroy(&da);
    return PetscFinalize();
}

PetscErrorCode formMatrix(DM da, Mat A) {
    PetscErrorCode ierr;
    DMDALocalInfo  info;
    MatStencil     row, col[5];
    PetscReal      hx, hy, v[5];
    PetscInt       i, j, ncols;

    ierr = DMDAGetLocalInfo(da,&info); CHKERRQ(ierr);
    hx = 1.0/(info.mx-1);  hy = 1.0/(info.my-1);
    for (j = info.ys; j < info.ys+info.ym; j++) {
        for (i = info.xs; i < info.xs+info.xm; i++) {
            row.j = j;           // row of A corresponding to (x_i,y_j)
            row.i = i;
            col[0].j = j;        // diagonal entry
            col[0].i = i;
            ncols = 1;
            if (i==0 || i==info.mx-1 || j==0 || j==info.my-1) {
                v[0] = 1.0;      // on boundary: trivial equation
            } else {
                v[0] = 2*(hy/hx + hx/hy); // interior: build a row
                if (i-1 > 0) {
                    col[ncols].j = j;    col[ncols].i = i-1;
                    v[ncols++] = -hy/hx;
                }
                if (i+1 < info.mx-1) {
                    col[ncols].j = j;    col[ncols].i = i+1;
                    v[ncols++] = -hy/hx;
                }
                if (j-1 > 0) {
                    col[ncols].j = j-1;  col[ncols].i = i;
                    v[ncols++] = -hx/hy;
                }
                if (j+1 < info.my-1) {
                    col[ncols].j = j+1;  col[ncols].i = i;
                    v[ncols++] = -hx/hy;
                }
            }
            ierr = MatSetValuesStencil(A,1,&row,ncols,col,v,INSERT_VALUES); CHKERRQ(ierr);
        }
    }
    ierr = MatAssemblyBegin(A,MAT_FINAL_ASSEMBLY); CHKERRQ(ierr);
    ierr = MatAssemblyEnd(A,MAT_FINAL_ASSEMBLY); CHKERRQ(ierr);
    return 0;
}


PetscErrorCode formExact(DM da, Vec uexact) {
    PetscErrorCode ierr;
    PetscInt       i, j;
    PetscReal      hx, hy, x, y, **auexact;
    DMDALocalInfo  info;

    ierr = DMDAGetLocalInfo(da,&info); CHKERRQ(ierr);
    hx = 1.0/(info.mx-1);  hy = 1.0/(info.my-1);
    ierr = DMDAVecGetArray(da, uexact, &auexact);CHKERRQ(ierr);
    for (j = info.ys; j < info.ys+info.ym; j++) {
        y = j * hy;
        for (i = info.xs; i < info.xs+info.xm; i++) {
            x = i * hx;
            auexact[j][i] = x*x * (1.0 - x*x) * y*y * (y*y - 1.0);
        }
    }
    ierr = DMDAVecRestoreArray(da, uexact, &auexact);CHKERRQ(ierr);
    return 0;
}

PetscErrorCode formRHS(DM da, Vec b) {
    PetscErrorCode ierr;
    PetscInt       i, j;
    PetscReal      hx, hy, x, y, f, **ab;
    DMDALocalInfo  info;

    ierr = DMDAGetLocalInfo(da,&info); CHKERRQ(ierr);
    hx = 1.0/(info.mx-1);  hy = 1.0/(info.my-1);
    ierr = DMDAVecGetArray(da, b, &ab);CHKERRQ(ierr);
    for (j=info.ys; j<info.ys+info.ym; j++) {
        y = j * hy;
        for (i=info.xs; i<info.xs+info.xm; i++) {
            x = i * hx;
            if (i==0 || i==info.mx-1 || j==0 || j==info.my-1) {
                ab[j][i] = 0.0;  // on boundary: 1*u = 0
            } else {
                f = 2.0 * ( (1.0 - 6.0*x*x) * y*y * (1.0 - y*y)
                    + (1.0 - 6.0*y*y) * x*x * (1.0 - x*x) );
                ab[j][i] = hx * hy * f;
            }
        }
    }
    ierr = DMDAVecRestoreArray(da, b, &ab); CHKERRQ(ierr);
    return 0;
}

